pkg/cloudprovider/vsphere/cloud.go

```diff
@@ -68,27 +67,17 @@ func (vs *VSphere) Initialize(clientBuilder controller.ControllerClientBuilder)
 
 		informerFactory := informers.NewSharedInformerFactory(client, controller.NoResyncPeriodFunc())
 
+		var connMgr *cm.ConnectionManager
 		if vs.cfg.Global.SecretNamespace != "" && vs.cfg.Global.SecretName != "" {
+			glog.V(4).Info("NewConnectionManagerK8s")
 			secretInformer := informerFactory.Core().V1().Secrets()
-			vs.nodeManager.credentialManager = &cm.SecretCredentialManager{
-				SecretName:            vs.cfg.Global.SecretName,
-				SecretNamespace:       vs.cfg.Global.SecretNamespace,
-				SecretLister:          secretInformer.Lister(),
-				SecretsDirectory:      vs.cfg.Global.SecretsDirectory,
-				SecretsDirectoryParse: false,
-				Cache: &cm.SecretCache{
-					VirtualCenter: make(map[string]*cm.Credential),
-				},
-			}
+			connMgr = cm.NewConnectionManagerK8s(vs.cfg, secretInformer.Lister())
 		} else if vs.cfg.Global.SecretsDirectory != "" {
-			vs.nodeManager.credentialManager = &cm.SecretCredentialManager{
-				SecretsDirectory:      vs.cfg.Global.SecretsDirectory,
-				SecretsDirectoryParse: false,
-				Cache: &cm.SecretCache{
-					VirtualCenter: make(map[string]*cm.Credential),
-				},
-			}
+			glog.V(4).Info("NewConnectionManagerGeneric")
+			connMgr = cm.NewConnectionManagerGeneric(vs.cfg)
 		}
+		vs.connectionManager = connMgr
+		vs.nodeManager.connectionManager = connMgr
```

Comment:
It doesn't have to be... but if some other part of the CCM needs to make API calls, the top level VSphere object makes it a convenient place to pass it from.

Reply:
yeah, I guess I am looking for any example of where we would pass `vs.nodeManager` but not `vs` itself, meaning can we skip the need for duplicating it in `vs.nodeManager` entirely? I expect the answer is no, but just checking.

Reply:
Yea, I would need to look into that. Just thinking about it really quickly... I suspect not because nodemanager functions by using cloud provider call backs to add/delete nodes. However, that is the reason why I created the CCM API... the only function on it right now is ListNodes() which gets you the info without having 2 sets of truth out there.

Reply:
I am thinking about this a little more and digesting what you were talking about... you definitely don't need to create `vs` type object if you don't need/want too.

As for nodemanager, really the only piece of code you need is the part where we attempt to discover which vCenter/datacenter combo the node (aka VM) belongs to. You might need to duplicate (or lift) that code and it's mainly because CSI should solve this "discovery" independent of the CO... ie you wouldn't be able to use the k8s callback that nodemanager relies on for DCOS.